src/EditorFeatures/Core/Implementation/IntelliSense/Completion/FileSystem/FileSystemCompletionHelper.cs

```diff
@@ -6,152 +6,211 @@
 using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
-using Microsoft.CodeAnalysis.Completion;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Shared.Utilities;
-using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.FileSystem
+namespace Microsoft.CodeAnalysis.Completion
 {
-    internal sealed class FileSystemCompletionHelper
+    internal class FileSystemCompletionHelper
     {
-        private readonly ICurrentWorkingDirectoryDiscoveryService _fileSystemDiscoveryService;
-        private readonly Func<string, bool> _exclude;
+        private static char[] s_windowsDirectorySeparator = { '\\' };
+
         private readonly Glyph _folderGlyph;
         private readonly Glyph _fileGlyph;
 
         // absolute paths
         private readonly ImmutableArray<string> _searchPaths;
+        private readonly string _baseDirectoryOpt;
 
-        private readonly ISet<string> _allowableExtensions;
+        private readonly ImmutableArray<string> _allowableExtensions;
 
         private readonly Lazy<string[]> _lazyGetDrives;
-        private readonly CompletionProvider _completionProvider;
-        private readonly TextSpan _textChangeSpan;
         private readonly CompletionItemRules _itemRules;
 
         public FileSystemCompletionHelper(
-            CompletionProvider completionProvider,
-            TextSpan textChangeSpan,
-            ICurrentWorkingDirectoryDiscoveryService fileSystemDiscoveryService,
             Glyph folderGlyph,
             Glyph fileGlyph,
             ImmutableArray<string> searchPaths,
-            IEnumerable<string> allowableExtensions,
-            Func<string, bool> exclude = null,
-            CompletionItemRules itemRules = null)
+            string baseDirectoryOpt,
+            ImmutableArray<string> allowableExtensions,
+            CompletionItemRules itemRules)
         {
             Debug.Assert(searchPaths.All(path => PathUtilities.IsAbsolute(path)));
+            Debug.Assert(baseDirectoryOpt == null || PathUtilities.IsAbsolute(baseDirectoryOpt));
 
-            _completionProvider = completionProvider;
-            _textChangeSpan = textChangeSpan;
             _searchPaths = searchPaths;
-            _allowableExtensions = allowableExtensions.Select(e => e.ToLowerInvariant()).ToSet();
-            _fileSystemDiscoveryService = fileSystemDiscoveryService;
+            _baseDirectoryOpt = baseDirectoryOpt;
+            _allowableExtensions = allowableExtensions;
             _folderGlyph = folderGlyph;
             _fileGlyph = fileGlyph;
-            _exclude = exclude;
             _itemRules = itemRules;
 
-            _lazyGetDrives = new Lazy<string[]>(() =>
-                IOUtilities.PerformIO(Directory.GetLogicalDrives, Array.Empty<string>()));
+            _lazyGetDrives = new Lazy<string[]>(GetLogicalDrives);
```

Comment:
Good point. Not sure if it's worth it to ask for the drives every time though. 

---
In reply to: [114853191](https://github.com/dotnet/roslyn/pull/19195#discussion_r114853191) [](ancestors = 114853191)